File: android-app/app/src/main/java/com/bimoraai/brahm/api/ApiService.java
package com.bimoraai.brahm.api;

import com.bimoraai.brahm.api.models.KundaliRequest;
import com.bimoraai.brahm.api.models.SendOtpRequest;
import com.bimoraai.brahm.api.models.SendOtpResponse;
import com.bimoraai.brahm.api.models.VerifyOtpRequest;
import com.bimoraai.brahm.api.models.VerifyOtpResponse;
import com.google.gson.JsonObject;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

/**
 * Retrofit interface for all Brahm AI backend endpoints.
 *
 * Base URL: https://brahmasmi.bimoraai.com/
 *
 * All endpoints that return variable / schema-free JSON use {@link JsonObject}
 * so callers can navigate the response freely without needing a separate POJO
 * for every endpoint.
 *
 * Auth / OTP endpoints use typed request/response POJOs for compile-time safety.
 */
public interface ApiService {

    // ─────────────────────────────────────────────────────────────────────────
    // Auth
    // ─────────────────────────────────────────────────────────────────────────

    /** Step 1 — send a one-time password to the given phone number. */
    @POST("api/auth/send-otp")
    Call<SendOtpResponse> sendOtp(@Body SendOtpRequest req);

    /** Step 2 — verify OTP and receive a JWT access token. */
    @POST("api/auth/verify-otp")
    Call<VerifyOtpResponse> verifyOtp(@Body VerifyOtpRequest req);

    // ─────────────────────────────────────────────────────────────────────────
    // Kundali
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Generate a full Drik-level Kundali for the given birth data.
     * Response contains: grahas, lagna, bhav_chalit, shadbala,
     * ashtakavarga, upagraha, yogas, dashas, alerts.
     */
    @POST("api/kundali")
    Call<JsonObject> getKundali(@Body KundaliRequest req);

    // ─────────────────────────────────────────────────────────────────────────
    // Panchang
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Daily Panchang for the given coordinates and date.
     *
     * @param lat  Latitude  (decimal degrees)
     * @param lon  Longitude (decimal degrees)
     * @param tz   UTC offset in hours (e.g. 5.5 for IST)
     * @param date ISO date string "YYYY-MM-DD"
     */
    @GET("api/panchang")
    Call<JsonObject> getPanchang(
        @Query("lat")  double lat,
        @Query("lon")  double lon,
        @Query("tz")   double tz,
        @Query("date") String date
    );

    // ─────────────────────────────────────────────────────────────────────────
    // Gochar (Transit)
    // ─────────────────────────────────────────────────────────────────────────

    /** Current planetary positions (transit sky chart). */
    @GET("api/gochar")
    Call<JsonObject> getGochar(
        @Query("lat") double lat,
        @Query("lon") double lon,
        @Query("tz")  double tz
    );

    /**
     * Analyse transits against the user's natal chart.
     * Body should contain birth_date, birth_time, lat, lon, tz.
     */
    @POST("api/gochar/analyze")
    Call<JsonObject> analyzeGochar(@Body JsonObject body);

    // ─────────────────────────────────────────────────────────────────────────
    // Horoscope
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Daily horoscope for a rashi.
     *
     * @param rashi Rashi slug e.g. "mesh", "vrishabha" (lowercase, no spaces).
     */
    @GET("api/horoscope/{rashi}")
    Call<JsonObject> getHoroscope(@Path("rashi") String rashi);

    // ─────────────────────────────────────────────────────────────────────────
    // Compatibility (Kundali Milan)
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Guna Milan / Nakshatra compatibility between two people.
     * Body: { person1: {date,time,lat,lon,tz}, person2: {date,time,lat,lon,tz} }
     */
    @POST("api/compatibility")
    Call<JsonObject> getCompatibility(@Body JsonObject body);

    // ─────────────────────────────────────────────────────────────────────────
    // Muhurta
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Find auspicious timings for a given activity.
     * Body: { activity, lat, lon, tz, date_from, date_to }
     */
    @POST("api/muhurta/activity")
    Call<JsonObject> getMuhurta(@Body JsonObject body);

    // ─────────────────────────────────────────────────────────────────────────
    // Sade Sati
    // ─────────────────────────────────────────────────────────────────────────

    @GET("api/sade-sati")
    Call<JsonObject> getSadeSati(
        @Query("lat")  double lat,
        @Query("lon")  double lon,
        @Query("tz")   double tz,
        @Query("date") String date,
        @Query("time") String time
    );

    // ─────────────────────────────────────────────────────────────────────────
    // Dosha
    // ─────────────────────────────────────────────────────────────────────────

    @GET("api/dosha")
    Call<JsonObject> getDosha(
        @Query("lat")  double lat,
        @Query("lon")  double lon,
        @Query("tz")   double tz,
        @Query("date") String date,
        @Query("time") String time
    );

    // ─────────────────────────────────────────────────────────────────────────
    // Gemstones
    // ─────────────────────────────────────────────────────────────────────────

    @GET("api/gemstones")
    Call<JsonObject> getGemstones(
        @Query("lat")  double lat,
        @Query("lon")  double lon,
        @Query("tz")   double tz,
        @Query("date") String date,
        @Query("time") String time
    );

    // ─────────────────────────────────────────────────────────────────────────
    // KP System
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * KP sub-lord analysis.
     * Body: { date, time, lat, lon, tz, name }
     */
    @POST("api/kp")
    Call<JsonObject> getKP(@Body JsonObject body);

    // ─────────────────────────────────────────────────────────────────────────
    // Prashna (Horary)
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Prashna Kundali for the current moment.
     * Body: { question, date, time, lat, lon, tz }
     */
    @POST("api/prashna")
    Call<JsonObject> getPrashna(@Body JsonObject body);

    // ─────────────────────────────────────────────────────────────────────────
    // Varshphal (Solar Return)
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Solar Return chart for a given target year.
     * Body: { date, time, lat, lon, tz, year }
     */
    @POST("api/varshphal")
    Call<JsonObject> getVarshphal(@Body JsonObject body);

    // ─────────────────────────────────────────────────────────────────────────
    // Birth Time Rectification
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * Rectify birth time using life events.
     * Body: { date, approx_time_from, approx_time_to, lat, lon, tz, life_events }
     */
    @POST("api/rectification")
    Call<JsonObject> getRectification(@Body JsonObject body);

    // ─────────────────────────────────────────────────────────────────────────
    // City search
    // ─────────────────────────────────────────────────────────────────────────

    /**
     * City autocomplete.
     *
     * @param query Partial city name (at least 2 characters).
     * Response: { cities: [{name, lat, lon, tz, country}] }
     */
    @GET("api/cities")
    Call<JsonObject> searchCities(@Query("q") String query);

    // ─────────────────────────────────────────────────────────────────────────
    // Sky / Planets now
    // ─────────────────────────────────────────────────────────────────────────

    /** Current positions of all planets in the sky. */
    @GET("api/planets/now")
    Call<JsonObject> getPlanetsNow(
        @Query("lat") double lat,
        @Query("lon") double lon,
        @Query("tz")  double tz
    );
}

File: android-app/app/src/main/kotlin/com/bimoraai/brahm/core/network/ApiService.kt
package com.bimoraai.brahm.core.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import retrofit2.Response
import retrofit2.http.*

// ─── Request / Response models ───────────────────────────────────────────────

@Serializable data class SendOtpRequest(val phone: String, val purpose: String = "login")
@Serializable data class SendOtpResponse(
    val sent: Boolean,
    val message: String,
    val test_otp: String? = null,   // returned in TEST_MODE
)

@Serializable data class VerifyOtpRequest(val phone: String, val otp: String, val purpose: String = "login")
@Serializable data class AuthResponse(
    val access_token: String,
    val refresh_token: String,
    val user_id: String,
    val name: String? = null,
    val plan: String = "free",
    val phone: String? = null,
    val phone_verified: Boolean = false,
)

@Serializable data class GoogleAuthRequest(val id_token: String, val device_type: String = "android")

// Used by GET /user — full profile
@Serializable data class UserDto(
    val session_id: String = "",
    val name: String = "",
    val phone: String? = null,
    val email: String? = null,
    val plan: String = "free",
    val date: String = "",        // birth_date YYYY-MM-DD
    val time: String = "",        // birth_time HH:MM
    val place: String = "",       // birth_place
    val gender: String = "",      // Male | Female | Other | Prefer not to say
    val lat: Double = 0.0,
    val lon: Double = 0.0,
    val tz: Double = 5.5,
    val rashi: String = "",
    val nakshatra: String = "",
)

// Used by POST /user — update profile
@Serializable data class UpdateProfileRequest(
    val session_id: String = "",
    val name: String = "",
    val date: String = "",
    val time: String = "",
    val lat: Double = 0.0,
    val lon: Double = 0.0,
    val tz: Double = 5.5,
    val place: String = "",
    val gender: String = "",
    val rashi: String = "",
    val nakshatra: String = "",
    val language: String = "english",
    val plan: String = "free",
    val phone: String? = null,
    val email: String? = null,
)

// PanchangRequest removed — backend uses GET with query params, not POST body

@Serializable data class KundaliRequest(
    val name: String = "",
    val date: String,            // YYYY-MM-DD
    val time: String,            // HH:MM
    val place: String = "",
    val lat: Double,
    val lon: Double,
    val tz: Double = 5.5,
    val ayanamsha: String = "lahiri",
    val rahu_mode: String = "mean",
    val calc_options: List<String> = listOf(
        "antardasha", "pratyantar", "sukshma",
        "ashtakavarga", "shadbala", "upagraha",
        // Varga charts — D-1 and D-9 always included in main response
        "d2", "d3", "d4", "d5", "d6", "d7", "d8",
        "d10", "d11", "d12", "d16", "d20", "d24", "d27",
        "d30", "d40", "d45", "d60",
    ),
)

@Serializable data class SaveKundaliRequest(
    val name: String = "",
    val birth_date: String,
    val birth_time: String,
    val birth_lat: Double,
    val birth_lon: Double,
    val birth_tz: Double = 5.5,
    val birth_place: String = "",
    val kundali_json: String,
)

@Serializable data class ChatRequest(
    val message: String,
    val conversation_id: String? = null,
)

@Serializable data class RefreshRequest(val refresh_token: String)

// City search
@Serializable data class City(
    val name: String,
    val label: String = "",
    val country: String = "",
    val lat: Double,
    val lon: Double,
    val tz: Double = 5.5,
)
@Serializable data class CitiesResponse(val cities: List<City>)
@Serializable data class CitySearchResponse(val results: List<City>)

// ─── API endpoints ────────────────────────────────────────────────────────────

interface ApiService {

    // Auth
    @POST("auth/otp/send")
    suspend fun sendOtp(@Body body: SendOtpRequest): Response<SendOtpResponse>

    @POST("auth/otp/verify")
    suspend fun verifyOtp(@Body body: VerifyOtpRequest): Response<AuthResponse>

    @POST("auth/google")
    suspend fun googleAuth(@Body body: GoogleAuthRequest): Response<AuthResponse>

    @POST("auth/refresh")
    suspend fun refreshToken(@Body body: RefreshRequest): Response<AuthResponse>

    @POST("auth/logout")
    suspend fun logout(): Response<Unit>

    // Cities — use search endpoint (GeoNames 200K+ DB) instead of legacy 730-city JSON
    @GET("cities/search")
    suspend fun searchCities(@Query("q") query: String, @Query("limit") limit: Int = 8): Response<CitySearchResponse>

    @GET("geocode")
    suspend fun geocode(@Query("q") query: String): Response<City>

    // User profile
    @GET("user")
    suspend fun getMe(): Response<UserDto>

    // Chat history sessions
    @GET("user/chats/sessions")
    suspend fun getChatSessions(
        @Query("include_archived") includeArchived: Boolean = false,
    ): Response<JsonObject>

    @DELETE("user/chats/session/{sessId}")
    suspend fun deleteChatSession(@Path("sessId") sessId: String): Response<JsonObject>

    @PATCH("user/chats/session/{sessId}/meta")
    suspend fun patchSessionMeta(
        @Path("sessId") sessId: String,
        @Body body: JsonObject,
    ): Response<JsonObject>

    @POST("user")
    suspend fun updateProfile(@Body body: UpdateProfileRequest): Response<UserDto>

    // Panchang — GET with query params (backend: GET /api/panchang?date=&lat=&lon=&tz=)
    @GET("panchang")
    suspend fun getPanchang(
        @Query("date") date: String,
        @Query("lat")  lat: Double,
        @Query("lon")  lon: Double,
        @Query("tz")   tz: Double = 5.5,   // IST offset in hours (NOT "Asia/Kolkata")
    ): Response<JsonObject>

    // Kundali
    @POST("kundali")
    suspend fun generateKundali(@Body body: KundaliRequest): Response<JsonObject>

    @GET("user/kundali")
    suspend fun getSavedKundali(): Response<JsonObject>

    @POST("user/kundali")
    suspend fun saveKundali(@Body body: SaveKundaliRequest): Response<JsonObject>

    // Gochar — current sky (GET, no body)
    @GET("gochar")
    suspend fun getGocharNow(): Response<JsonObject>

    // All other astrology endpoints — POST with dynamic bodies
    @POST("compatibility")
    suspend fun getCompatibility(@Body body: JsonObject): Response<JsonObject>

    @POST("muhurta/activity")
    suspend fun getMuhurta(@Body body: JsonObject): Response<JsonObject>

    @POST("kp")
    suspend fun getKP(@Body body: JsonObject): Response<JsonObject>

    @POST("prashna")
    suspend fun getPrashna(@Body body: JsonObject): Response<JsonObject>

    @POST("varshphal")
    suspend fun getVarshphal(@Body body: JsonObject): Response<JsonObject>

    @POST("rectification")
    suspend fun getRectification(@Body body: JsonObject): Response<JsonObject>

    @GET("horoscope/{rashi}")
    suspend fun getHoroscope(
        @Path("rashi") rashi: String,
        @Query("period") period: String = "daily",
    ): Response<JsonObject>

    // Palmistry (multipart — image upload)
    @Multipart
    @POST("palmistry/analyze")
    suspend fun analyzePalm(
        @Part image: okhttp3.MultipartBody.Part,
        @Query("hand_role") handRole: String = "dominant",
    ): Response<JsonObject>

    @Multipart
    @POST("palmistry/analyze-both")
    suspend fun analyzeBothPalms(
        @Part dominant: okhttp3.MultipartBody.Part,
        @Part non_dominant: okhttp3.MultipartBody.Part,
        @Query("dominant_hand") dominantHand: String = "right",
    ): Response<JsonObject>

    // Sky — current planetary positions
    @GET("planets/now")
    suspend fun getPlanetsNow(): Response<JsonObject>

    // Vedic Library — RAG search
    @GET("search")
    suspend fun searchVedicLibrary(@Query("q") query: String): Response<JsonObject>

    // Festivals — full year (for PanchangScreen Festivals tab)
    @GET("festivals")
    suspend fun getFestivalsYear(
        @Query("year") year: Int,
        @Query("lat")  lat: Double,
        @Query("lon")  lon: Double,
        @Query("tz")   tz: Double = 5.5,
    ): Response<JsonObject>

    // Festivals — per month (legacy, kept for compatibility)
    @GET("festivals")
    suspend fun getFestivals(
        @Query("month") month: Int,
        @Query("year")  year: Int,
    ): Response<JsonObject>

    // Grahan / Eclipse calendar (for PanchangScreen)
    @GET("grahan")
    suspend fun getGrahan(): Response<JsonObject>

    // Gochar personal analysis
    @POST("gochar/analyze")
    suspend fun analyzeGochar(@Body body: JsonObject): Response<JsonObject>

    // Calendar — monthly grid with per-day tithi, festivals, special day flags
    @GET("calendar/month")
    suspend fun getCalendar(
        @Query("year")         year: Int,
        @Query("month")        month: Int,
        @Query("lat")          lat: Double,
        @Query("lon")          lon: Double,
        @Query("tz")           tz: Double = 5.5,
        @Query("tradition")    tradition: String = "smarta",
        @Query("lunar_system") lunarSystem: String = "amanta",
    ): Response<JsonObject>
}
